EELSModel.estimate_and_set_coreloss_edge_onset

magnunor opened the pull request:
As discussed a long time ago in https://github.com/hyperspy/hyperspy/issues/487, there is currently no good way of setting the onsets of the `EELSCLEdge` components.

This pull requests implements the method of doing this as outlined in https://doi.org/10.1016/j.ultramic.2016.07.004 (arXiv: https://arxiv.org/abs/1608.07814).

The method itself is explained here: https://github.com/hyperspy/hyperspy/pull/490#issue-32186703

### Changes

- Adds `misc.eels.tools.get_edge_onset`: to find onset using the method explained above
- Adds `misc.utils.find_nearest_index` and `misc.utils.find_nearest_index_from_right`: which are utility functions for `misc.eels.tools.get_edge_onset`
- `EELSModel.set_coreloss_edge_onset`: which is the main function people will use, which calls either the UI function in `signal_tools` or `EELSModel._set_coreloss_edge_onset`.
- `signal_tools.SetCorelossEdgeOnset`: which handles the UI parts of the implementation
- New widget in `hyperspy_gui_traitsui`: https://github.com/hyperspy/hyperspy_gui_traitsui/pull/12
- New widget in `hyperspy_gui_ipywidgets`: https://github.com/hyperspy/hyperspy_gui_ipywidgets/pull/7

### Progress of the PR
- [x] Implemenation
- [x] Add docstrings
- [x] update user guide
- [x] add tests
- [x] ready for review

### Minimal example of the bug fix or the new feature
```python
import hyperspy.api as hs
s = hs.datasets.artificial_data.get_core_loss_eels_line_scan_signal()
s.set_microscope_parameters(
        beam_energy=200, convergence_angle=26, collection_angle=20)
s.add_elements(['Mn', ])
m = s.create_model(auto_background=False, GOS='hydrogenic')
m.set_coreloss_edge_onset(m.components.Mn_L3)
```

![set_coreloss_edge_onset](https://user-images.githubusercontent.com/1690979/43369817-a939d91a-9374-11e8-8251-1a11ff7a6c66.png)

Old pull requests: https://github.com/hyperspy/hyperspy/pull/1767, https://github.com/hyperspy/hyperspy/pull/1271, https://github.com/hyperspy/hyperspy/pull/490

magnunor committed:
NEW: function for setting onset energy of EELSCLEdge components

magnunor committed:
Remove unneeded import

magnunor committed:
m.set_coreloss_edge_onset was using the wrong call non-interactive use

magnunor committed:
Add tests for m.set_coreloss_edge_onset

magnunor committed:
DOC: add small description about m.set_coreloss_edge_onset in user guide

magnunor committed:
Improve utils.misc.find_nearest_index_from_right: use numpy operations

Instead of using a for-loop. Also change how threshold is used.

magnunor committed:
Improve docstring for eelsmodel.set_coreloss_edge_onset

magnunor committed:
Fix typo in eelsmodel.set_coreloss_edge_onset docstring

Commit:
Merge branch 'master' into NEW_get_edge_onset

Conflicts:
	hyperspy/misc/utils.py
	hyperspy/models/eelsmodel.py

magnunor committed:
Cleanup: set_coreloss_edge_onset and _set_coreloss_edge_onset function

magnunor committed:
Merge branch 'RELEASE_next_minor' into NEW_get_edge_onset

magnunor committed:
Cleanup of get_edge_onset and update set_coreloss_edge_onset tests

magnunor committed:
Update set_coreloss_edge_onset + its unit tests

magnunor committed:
Remove interactive parts of set_coreloss_edge_onset

magnunor committed:
Merge branch 'RELEASE_next_minor' into NEW_set_edge_onset_cli

magnunor committed:
Move get_edge_onset to misc/eels/tools

magnunor committed:
unit tests: flake8 fixes in misc/test_utils

Also change imports in that test, to allow for more tests

magnunor committed:
Improve find_nearest_index_from_right: threshold check, docstring, tests

magnunor committed:
Improve get_edge_onset, and add unit tests

magnunor committed:
get_edge_onset: add check for percent_position input

Also add unit test

magnunor committed:
Improve EELSModel.set_edge_onset and add unit tests

magnunor committed:
Add GUI tools for EELSModel.set_coreloss_edge_onset

magnunor committed:
Add reference to article detailing set_coreloss_edge_onset

francisco-dlp commented on hyperspy/misc/eels/tools.py:
```diff
@@ -307,3 +309,63 @@ def eels_constant(s, zlp, t):
+        raise ValueError(
+                "percent position ({0}) must be between 0 and 1".format(
+                    percent_position))
+    start_i = utils.find_nearest_index(x_axis, start)
```
Why not using ``value2index`` instead? Avoiding adding extra functions whenever possible should help easing the task of maintaining the code.

magnunor replied:
This doesn't work on `EELSSpectrum` but directly on numpy arrays, so `value2index` can't be called here. The reasoning for that is trying to have a bit cleaner separation between the HyperSpy signal parts, and the "calculation" parts.

francisco-dlp commented on hyperspy/misc/eels/tools.py:
```diff
@@ -307,3 +309,63 @@ def eels_constant(s, zlp, t):
+                "percent position ({0}) must be between 0 and 1".format(
+                    percent_position))
+    start_i = utils.find_nearest_index(x_axis, start)
+    end_i = utils.find_nearest_index(x_axis, end)
```
As above.

francisco-dlp commented on hyperspy/misc/eels/tools.py:
```diff
@@ -307,3 +309,63 @@ def eels_constant(s, zlp, t):
+    data_onset_value = (data_max - data_min) * percent_position + data_min
+
+    interpolate = interp1d(x_axis, data, fill_value='extrapolate')
+    x_axis_interp = np.arange(x_axis[0], x_axis[-1], 0.0001)
```
Wouldn't it be possible to use ``spline.roots`` for this instead of computing the interpolated spectrum with the step hardwired?

Beside, what's the point of using interpolation to estimate the onset position using a method that is only qualitative?

I suggest to make interpolation optional. This would have the advantage that the algorithm could be vectorized boosting performance.

magnunor replied:
This is greatly improved now, so no interpolation is done. A spline is used to get subpixel accuracy between two values.

francisco-dlp commented on hyperspy/misc/eels/tools.py:
```diff
@@ -307,3 +309,63 @@ def eels_constant(s, zlp, t):
+        data_min_i = temp_i
+
+    threshold = 0.0001 * (data_max - data_min)
+    data_onset_i = utils.find_nearest_index_from_right(
```
isn't it possible to use ``value2index`` here too?

francisco-dlp commented on hyperspy/models/eelsmodel.py:
```diff
@@ -976,3 +978,82 @@ def resume_auto_fine_structure_width(self, update=True):
         else:
             warnings.warn("Not suspended, nothing to resume.")
+
+    def set_coreloss_edge_onset(
```
What about calling this ``estimate_and_set_coreloss_edge_onset``?

magnunor replied:
Done!

francisco-dlp reviewed (changes requested):
Nice and polished contribution!

magnunor committed:
Remove find_nearest_index, find_nearest_index_from_right in misc.utils

The functions were used in eels.tools.get_edge_onset, but is no longer
necessary.

magnunor committed:
Optimized implementation of eels.tools.get_edge_onset

Now it avoids interpolating the whole signal

magnunor committed:
Remove tests for function which was previously removed

utils.find_nearest_index and utils.find_nearest_index_from_right

magnunor committed:
Flake8 fixes

magnunor committed:
Delete file with tests which have been moved to test_eelsmodel.py

magnunor committed:
Rename set_coreloss_edge_onset to estimate_and_set_coreloss_edge_onset

magnunor committed:
Fix wrong function call in estimate_and_set_coreloss_edge_onset

magnunor committed:
Minor fix to eels.tools.get_edge_onset

magnunor committed:
docstring: fix function reference

magnunor commented:
I reimplemented most of the `get_edge_onset` function, so it no longer interpolates the whole spectrum. Now it finds the index in the original spectrum, and uses a spline between that index and the next, to get subpixel accuracy. The `find_index...` functions were also removed, since they were no longer necessary.

francisco-dlp commented on hyperspy/misc/eels/tools.py:
```diff
@@ -307,3 +308,61 @@ def eels_constant(s, zlp, t):
+                    percent_position))
+    start_i = np.argmax(x_axis > start)
+    end_i = np.argmax(x_axis > end)
+    if end_i == 0:
```
Woulndn't it be better to raise a ``ValueError`` here?

magnunor replied:
Fixed by adding several ifs at the beginning of the function, to check if start/end/+++ is out of bounds of x-axis.

francisco-dlp commented on hyperspy/misc/eels/tools.py:
```diff
@@ -307,3 +308,61 @@ def eels_constant(s, zlp, t):
+    data_max = data.max()
+    data_min = data.min()
+    data_max_i = data.argmax()
+    data_min_i = data.argmin()
```
I would add here a sanity check: ``if data_max_i < data_min_i``. This may happen when analysing multi-dimensional datasets if the edge is not present at a given location. Actually, this sanity check could be used to assign a ``nan`` for the onset at those pixels.

magnunor replied:
Good idea! However, I'm not sure how to add a `np.nan` to a `EELSCLEdge` component.
```python
Mn_L3.onset_energy.value = np.nan
```
Gives the error
```python
hyperspy/hyperspy/misc/eels/base_gos.py in get_qaxis_and_gos(self, ienergy, qmin, qmax)
     57         else:
     58             index = self.qaxis.searchsorted(qmax)
---> 59             g1, g2 = qgosi[index - 1:index + 1]
     60             q1, q2 = self.qaxis[index - 1: index + 1]
     61             gosqmax = get_linear_interpolation((q1, g1), (q2, g2), qmax)

ValueError: not enough values to unpack (expected 2, got 1)
```

francisco-dlp commented on hyperspy/misc/eels/tools.py:
```diff
@@ -307,3 +308,61 @@ def eels_constant(s, zlp, t):
+    start_i = np.argmax(x_axis > start)
+    end_i = np.argmax(x_axis > end)
+    if end_i == 0:
+        end_i = len(x_axis)
```
Wouldn't it be better to raise a ``ValueError`` here?

magnunor replied:
Fixed by adding several `if`s at the beginning of the function, to check if `start`/`end`/+++ is out of bounds of x-axis.

francisco-dlp commented on hyperspy/misc/eels/tools.py:
```diff
@@ -307,3 +308,61 @@ def eels_constant(s, zlp, t):
+    onset_slice = slice(data_min_i + data_onset_i - 1,
+                        data_min_i + data_onset_i + 1)
+
+    spline = InterpolatedUnivariateSpline(
```
I still think that spline interpolation is an overkill feature here that breaks the "vectority" of the algorithm. Actually, here a spline is used just for linear interpolation. Why not writing the linear interpolator vectorially instead?

francisco-dlp commented on hyperspy/models/eelsmodel.py:
```diff
@@ -976,3 +978,83 @@ def resume_auto_fine_structure_width(self, update=True):
+                x_axis, percent_position=percent_position)
+            component.onset_energy.value = edge_onset_energy
+            component.onset_energy.store_current_value_in_array()
+        else:
```
If the algorithm is written in vectorial form, this iteration becomes unnecessary.

francisco-dlp commented on hyperspy/models/eelsmodel.py:
```diff
@@ -976,3 +978,83 @@ def resume_auto_fine_structure_width(self, update=True):
         else:
             warnings.warn("Not suspended, nothing to resume.")
+
+    def estimate_and_set_coreloss_edge_onset(
```
Wouldn't it be good to add a ``mask`` kwarg to mask positions where the edge is not present?

francisco-dlp reviewed (changes requested):
Other than the comments above, the User Guide / docstring suggestions from my previous review has not been addressed. A clear discussion of the limits of the method could prevent misuse by people who may otherwise blindly believe the values returned by the algorithm.

magnunor commented:
For some reason your previous comments about user guide/docstrings changes doesn't show up for me.

Was this the comment about this?

>  A clear discussion of the limits of the method could prevent misuse by people who may otherwise blindly believe the values returned by the algorithm.

magnunor committed:
get_edge_onset: add value checking of start and end parameters

magnunor committed:
get_edge_onset: remove spline interpolation

Also some work with vectorizing the function, but still not
complete. So currently it only works with a single spectra at a time

magnunor commented:
I replaced the spline with a simple linear calculation.

However, the `get_edge_onset` function still only works on single spectra, and I don't have time to improve that aspect of it before October.

magnunor committed:
Merge branch 'RELEASE_next_minor' into NEW_set_coreloss_edge_onset

magnunor committed:
Update estimate_and_set_coreloss_edge_onset to hyperspy_extensions

magnunor committed:
get_edge_onset: return np.nan if edge is not found

Note: this does not currently save it to EELSCLEdge.onset_energy.map

magnunor commented:
With the EELS functionality having been moved to eXSpy, I'm guessing this pull request should be moved there as well?

If yes, one question would be where the GUI functionality should be placed? Should they be in `hyperspy_gui_ipywidget` and `hyperspy_gui_traitsui`?

ericpre commented:
They can go in either: `eXSpy` or the `hyperspy_gui_*` - there are already quite a few in `hyperspy_gui_*`: https://github.com/hyperspy/hyperspy_gui_ipywidgets/blob/main/hyperspy_gui_ipywidgets/hyperspy_extension.yaml
Since the existing ones are already in `hyperspy_gui_*` it would make sense to keep following this pattern?
